### PR TITLE
Add mesos authentication to the mesos cli

### PR DESCRIPTION
**Pull request by andreaspeters**
The following points I done:

- Add authentication against mesos manager and agent
- Add option to skip SSL verification of the mesos-agent
- Changed the order of "task list" to get back more running states

### CODE REVIEW

File: src/python/cli_new/lib/cli/http.py
Comment (asekretenko) [inline src/python/cli_new/lib/cli/http.py:77]: Well, I understand that this code used to silently drop arbitrary errors since time immemorial, but probably now it is the right time to fix this:)

File: src/python/cli_new/lib/cli/mesos.py
Comment (asekretenko) [inline src/python/cli_new/lib/cli/mesos.py:519]: Given that this code occurs three times, it is definitely worth extracting into a dedicated function.

File: src/python/cli_new/lib/cli/mesos.py
Comment (asekretenko) [inline src/python/cli_new/lib/cli/mesos.py:561]: You can just wrap the line instead of suppressing pylint warnings:
```
if self.config.agent_principal() is not None and \
    self.config.agent_secret() is not None:
```

**Review by asekretenko — COMMENTED** [review]
The approach looks good overall; thanks for addressing this!

File: src/python/cli_new/lib/cli/config.py
Comment (andreaspeters) [reply src/python/cli_new/lib/cli/config.py:127]: I add the check if the parameter set and not empty. Do u have other checks in your mind? :-)

File: src/python/cli_new/lib/cli/mesos.py
Comment (andreaspeters) [reply src/python/cli_new/lib/cli/mesos.py:519]: U are right. I also moved it into the config.py. but I'm unsteady, maybe it makes more sense in http.py. :thinking:

File: src/python/cli_new/lib/cli/http.py
Comment (andreaspeters) [reply src/python/cli_new/lib/cli/http.py:54]: right, I move it into the try->exception! :-)

File: src/python/cli_new/lib/cli/http.py
Comment (andreaspeters) [reply src/python/cli_new/lib/cli/http.py:81]: Well, good that you speak it out. The loop also bother me. So I removed it and add a timeout into the read_endpoint function. :-) Hope it looks good now.

**Comment by andreaspeters** [comment]
Does we have a list of features we want to see in the mesos_cli? I'm also planning to develope a plugin for our kubernetes framework.

**Comment by andreaspeters** [comment]
@asekretenko I'm done now. U can review again. :-)

File: src/python/cli_new/lib/cli/config.py
Comment (cf-natali) [reply src/python/cli_new/lib/cli/config.py:128]: Expanding on @asekretenko suggestion, this might be better:
```
principal = self.data["master"].get("principal")
if principal is None:
    return None
elif not principal:
    raise CLIException("The 'principal' field in 'master'"
                                   " must be non-empty")
return principal
```

Same thing for the other accessors.

File: src/python/cli_new/lib/cli/http.py
Comment (cf-natali) [inline src/python/cli_new/lib/cli/http.py:36]: Looking at the code below it doesn't seem that the function can ever return `None` for `data` - maybe no need to set it to `None` here, and just directly return below:
```
return http_response.data.decode('utf-8')
```

File: src/python/cli_new/lib/cli/http.py
Comment (cf-natali) [inline src/python/cli_new/lib/cli/http.py:75]: Hm, I'm not sure I understand what's going on here - the function's docstring says "If we are unable to read the data we throw an error.", so why catch the exception?
And also, if we catch an exception here, `data` will be `None`, so the below `data = json.loads(data)` is definitely not going to work.

File: src/python/cli_new/lib/cli/mesos.py
Comment (cf-natali) [inline src/python/cli_new/lib/cli/mesos.py:116]: Maybe instead of disabling the pylint warning which is legitimate it'd be better to default `query=None` and in the function's body do:
```
if query is None:
    query = {'order':'asc'}
```

File: src/python/cli_new/lib/cli/mesos.py
Comment (cf-natali) [inline src/python/cli_new/lib/cli/mesos.py:202]: Might be more readable:
```
scheme = "https://" if config.agent_ssl() else "http://"
```

File: src/python/cli_new/lib/cli/http.py
Comment (andreaspeters) [reply src/python/cli_new/lib/cli/http.py:75]: Sorry this part I just removed the loop around. I don't know what the idea behind this exception handling was. But I think u make a good point. Let me have a look to see if I can change.

File: src/python/cli_new/lib/cli/mesos.py
Comment (andreaspeters) [reply src/python/cli_new/lib/cli/mesos.py:116]: done

File: src/python/cli_new/lib/cli/http.py
Comment (andreaspeters) [reply src/python/cli_new/lib/cli/http.py:36]: yes, u are right. :+1:

File: src/python/cli_new/lib/cli/mesos.py
Comment (andreaspeters) [reply src/python/cli_new/lib/cli/mesos.py:202]: Oh yes, that's pretty. Didn't know that. :-) Thanks @cf-natali

File: src/python/cli_new/lib/cli/http.py
Comment (andreaspeters) [reply src/python/cli_new/lib/cli/http.py]: So, I just remove the exception handling. Like you said, I make so sense there. :+1:

**Comment by kamaradclimber** [comment]
Could you squash all commits to increase readability of this history before merging?

**Comment by andreaspeters** [comment]
> Could you squash all commits to increase readability of this history before merging?
Done :-)

**Comment by kamaradclimber** [comment]
Are you sure? I still see 17 commits.

**Comment by andreaspeters** [comment]
> Are you sure? I still see 17 commits.

:thinking:  not 100%! But like u see, the last commit include all changes. :man_shrugging:

**Comment by andreaspeters** [comment]
> > Are you sure? I still see 17 commits.
> 
> thinking not 100%! But like u see, the last commit include all changes. man_shrugging

Yeah it looks like I done something wrong during my past commits that the squash didn't worked as I expected. Can u squash during the MR?

**Comment by cf-natali** [comment]
> > > Are you sure? I still see 17 commits.
> > 
> > 
> > thinking not 100%! But like u see, the last commit include all changes. man_shrugging
> 
> Yeah it looks like I done something wrong during my past commits that the squash didn't worked as I expected. Can u squash during the MR?

Maybe, but it'll be easier using `git rebase`. Then just need to `git push -f`

**Comment by andreaspeters** [comment]
> > > > Are you sure? I still see 17 commits.
> > > 
> > > 
> > > thinking not 100%! But like u see, the last commit include all changes. man_shrugging
> > 
> > 
> > Yeah it looks like I done something wrong during my past commits that the squash didn't worked as I expected. Can u squash during the MR?
> 
> Maybe, but it'll be easier using `git rebase`. Then just need to `git push -f`

Hopefully I didn't broke more as I fixed with the rebase. :sweat_smile:

File: src/python/cli_new/lib/cli/config.py
Comment (cf-natali) [reply src/python/cli_new/lib/cli/config.py:128]: Hm, you don't seem to have done this?

Comment (andreaspeters) [reply src/python/cli_new/lib/cli/config.py:128]: I didn't feel it's a so big different to before. Thats why I ask! But I have no problem to change it. What do u think?

Comment (andreaspeters) [reply src/python/cli_new/lib/cli/config.py]: Ok, I change it. :-) Shell I squash the commit again? :joy:

Comment (andreaspeters) [reply src/python/cli_new/lib/cli/config.py]: Actually, I can also remove the Exception handling. The principal, and secret is not a "must-have".

File: src/python/cli_new/lib/cli/config.py
Comment (cf-natali) [inline src/python/cli_new/lib/cli/config.py:158]:
```
isinstance(agent_ssl, bool)
```

File: src/python/cli_new/lib/cli/http.py
Comment (cf-natali) [inline src/python/cli_new/lib/cli/http.py:29]: Why is this necessary?

File: src/python/cli_new/lib/cli/http.py
Comment (cf-natali) [inline src/python/cli_new/lib/cli/http.py:52]: Maybe make `timeout` a parameter with default 5 instead of hard-coding it, so it could be specified if needed.

File: src/python/cli_new/lib/cli/config.py
Comment (andreaspeters) [reply src/python/cli_new/lib/cli/config.py:158]: Is already change in the current version.  But I'm interesting of what do u think about to change `def secret` and `def principal` to 

```
    def secret(self):
        return self.data["master"].get("secret")
```
No matter secret does not exist, or it's empty, secret would be "None". Thats why the "elif" is not usefull in these case. And because secret does not have to be configured, it's also not important to rise a "Exception".  What do u think?

File: src/python/cli_new/lib/cli/http.py
Comment (andreaspeters) [reply src/python/cli_new/lib/cli/http.py:29]: I'm getting a lot of non pretty warning if I'm using a self signed certificate. But because of the possibility to set "ssl_verify", didn't think it made sense to issue the warning as well.

File: src/python/cli_new/lib/cli/http.py
Comment (andreaspeters) [reply src/python/cli_new/lib/cli/http.py:52]: :+1: good idea. 
Even if it's not easy to see that someone "want" have so many changes in the code, I feel like the code is getting better and better. :-) The discuss about some points helping a lot. So, thank's @cf-natali @asekretenko @kamaradclimber

File: src/python/cli_new/lib/cli/config.py
Comment (cf-natali) [reply src/python/cli_new/lib/cli/config.py:158]: > Is already change in the current version.

I don't seem to see it.

Also, I still see 3 commits, I thought you'd squashed?

Comment (cf-natali) [reply src/python/cli_new/lib/cli/config.py:158]: > But I'm interesting of what do u think about to change `def secret` and `def principal` to
> 
> ```
>     def secret(self):
>         return self.data["master"].get("secret")
> ```
> 
> No matter secret does not exist, or it's empty, secret would be "None". Thats why the "elif" is not usefull in these case. And because secret does not have to be configured, it's also not important to rise a "Exception". What do u think?

Yes, if it's not required it makes sense.

File: src/python/cli_new/lib/cli/http.py
Comment (cf-natali) [reply src/python/cli_new/lib/cli/http.py:29]: Then maybe add a comment explaining it?

File: src/python/cli_new/lib/cli/config.py
Comment (andreaspeters) [reply src/python/cli_new/lib/cli/config.py:158]: :+1:

File: src/python/cli_new/lib/cli/http.py
Comment (andreaspeters) [reply src/python/cli_new/lib/cli/http.py:29]: Will do it :-)

File: src/python/cli_new/lib/cli/config.py
Comment (andreaspeters) [reply src/python/cli_new/lib/cli/config.py:158]: Yeah I didnt commit and pushed just now. I wanted to wait until I get the feedback to the "def secret". :-)  Then I will also squash again.

File: src/python/cli_new/lib/cli/config.py
Comment (cf-natali) [inline src/python/cli_new/lib/cli/config.py:174]: I think you can do instead:
```
            timeout = self.data["agent"].get("timeout", 5)
```

Also, maybe better than hard-coding the default could be moved to the function prototype?
```
    def agent_timeout(self, default=5):
```

File: src/python/cli_new/lib/cli/config.py
Comment (cf-natali) [inline src/python/cli_new/lib/cli/config.py:192]: This seems equivalent to just:
```
return self.data["agent"].get("principal")
```

File: src/python/cli_new/lib/cli/config.py
Comment (cf-natali) [inline src/python/cli_new/lib/cli/config.py:203]: This seems equivalent to just:
```
return self.data["agent"].get("secret")
```

File: src/python/cli_new/lib/cli/http.py
Comment (cf-natali) [inline src/python/cli_new/lib/cli/http.py:53]: I'm not familiar with `urllib3` - what does `Poolmanager` bring over just using  `utllib.request`?
Does it not support custom headers?

File: src/python/cli_new/lib/cli/mesos.py
Comment (cf-natali) [inline src/python/cli_new/lib/cli/mesos.py:420]: What happens if `elf.config.agent_ssl_verify()` is `None`, will it work?

File: src/python/cli_new/lib/cli/plugins/task/main.py
Comment (cf-natali) [inline src/python/cli_new/lib/cli/plugins/task/main.py:82]: Nit: I think the indentation with the newline before was more readable.

File: src/python/cli_new/lib/cli/config.py
Comment (andreaspeters) [reply src/python/cli_new/lib/cli/config.py:192]: :joy: I was not so brave. But of course, u are right.

File: src/python/cli_new/lib/cli/config.py
Comment (andreaspeters) [reply src/python/cli_new/lib/cli/config.py:174]: I like the idea with the default value in the prototype. I will use it also for the agent_ssl and agent_ssl_verify. That will also simplify that functions. :+1:  thanks Charles.

File: src/python/cli_new/lib/cli/mesos.py
Comment (andreaspeters) [reply src/python/cli_new/lib/cli/mesos.py:420]: Thats obsolete now :-)

**Comment by ridv** [comment]
> Could you squash all commits to increase readability of this history before merging?

In the past the commiter merging the change used the "squash and merge" strategy which has the same effect. So no need to do anything special  😄 . (In fact, it's probably better to leave the individual commits to see how the patch progressed which is a bit of a carry over from when reviews happened on Review Board)

File: src/python/cli_new/lib/cli/config.py
Comment (cf-natali) [inline src/python/cli_new/lib/cli/config.py:147]: So you can simplify it even further by using a default for `self.data["agent"], like this]:
```
def agent_ssl(self, default=False):
    agent_ssl = self.data.get("agent", {}).get("ssl", default)
    if not isinstance(agent_ssl, bool):
        raise CLIException("The 'agent->ssl' field must be True/False"
    return agent_ssl
```

Personally I like it because it removes a repetition between checking that `"agent"` is in `data` and then accessing it, and also it removes a branch, but up to you if you find it less natural.

File: src/python/cli_new/lib/cli/config.py
Comment (cf-natali) [inline src/python/cli_new/lib/cli/config.py:153]: Same here.

File: src/python/cli_new/lib/cli/config.py
Comment (cf-natali) [inline src/python/cli_new/lib/cli/config.py:220]: Nit: the indentation seems a bit strange?

File: src/python/cli_new/lib/cli/plugins/task/main.py
Comment (cf-natali) [inline src/python/cli_new/lib/cli/plugins/task/main.py:111]: Why store it here and not directly pass `self.config` below?
```
        try:
            tasks = get_tasks(master, config)
```

File: src/python/cli_new/lib/cli/config.py
Comment (andreaspeters) [reply src/python/cli_new/lib/cli/config.py:147]: I also was thinking to replace it like that, but in my eyes it looks too complicated. :thinking: You don't think so?

File: src/python/cli_new/lib/cli/config.py
Comment (andreaspeters) [reply src/python/cli_new/lib/cli/config.py:220]: The linter say it's fine like that. :man_shrugging:  And my vim say, it should be like that. :joy:

File: src/python/cli_new/lib/cli/plugins/task/main.py
Comment (andreaspeters) [reply src/python/cli_new/lib/cli/plugins/task/main.py:111]: I thinks it's easier to read.

Comment (cf-natali) [reply src/python/cli_new/lib/cli/plugins/task/main.py:111]: OK!

File: src/python/cli_new/lib/cli/config.py
Comment (cf-natali) [reply src/python/cli_new/lib/cli/config.py:147]: Up to you, I'm fine either way :).

**Review by cf-natali — APPROVED** [review]

**Comment by cf-natali** [comment]
@asekretenko @qianzhangxa 
Maybe you could have a look at it?

**Comment by andreaspeters** [comment]
> @asekretenko @qianzhangxa
> Maybe you could have a look at it?

Some feedback? :-)

**Comment by qianzhangxa** [comment]
I am not very familiar with Python :(
@asekretenko can you please review again?

File: src/python/cli_new/lib/cli/http.py
Comment (asekretenko) [inline src/python/cli_new/lib/cli/http.py:85]: Does this mean that this code now returns `data` anyway, regardless of whether `condition(data)` returned `true` or `false`? Is the `condition` parameter really needed then?

It occurs to me that we have a lot of code in the tests which used to use this function to wait until `condition` becomes true (example: https://github.com/apache/mesos/blob/72f16f68973bf7d2ce5c621539a21fc4eccfa56e/src/python/cli_new/lib/cli/tests/base.py#L280).

While removing a lame retry loop is a good thing (this loop in its existing form is harmful for production environments, as it contains no exponential backoff logic), something needs to be done with these tests make use of `condition`.

If these tests are dead code, that dead code should be removed (ideally, in a separate commit). 
On the other hand, if this wait-for-condition functionality is really needed by tests, I would say that it should be pulled into a separate function wrapping the new `get_json()`. Given that this is apparently only used in tests, this new wrapper fucnction should be a part of the tests suite and not of this file.

**Comment by asekretenko** [comment]
Other than this question about the `condition` logic, I don't see any other issues.

File: src/python/cli_new/lib/cli/http.py
Comment (andreaspeters) [reply src/python/cli_new/lib/cli/http.py:85]: Actually I'm not 100% sure what the idea behind `condition` was. But yes, `data` have ever be to returned. But if `condition` is set, then through that function. I will create (for myself) a issue to investigate `condition` to see if it's needed or if we can remove it. Then we can merge this PR. Is this procedure ok for everyone? :-)

Comment (andreaspeters) [reply src/python/cli_new/lib/cli/http.py:85]: Actually I'm not 100% sure what the idea behind condition was. But yes, data have ever be to returned. But if condition is set, then through that function. I will create (for myself) a issue to investigate condition to see if it's needed or if we can remove it. Then we can merge this PR. Is this procedure ok for everyone? :-)

**Comment by andreaspeters** [comment]
Can we merge it please?

**Comment by qianzhangxa** [comment]
Sorry for the late response, I have merged it, thanks for the contributions @andreaspeters!

**Comment by andreaspeters** [comment]
No problem @qianzhangxa. Your welcome. :-)

**Pull request closed without merging**
